src/Processors/Formats/Impl/NativeORCBlockInputFormat.cpp

```diff
@@ -946,24 +981,68 @@ void NativeORCBlockInputFormat::prepareFileReader()
     }
     include_indices.assign(include_typeids.begin(), include_typeids.end());
 
-    if (format_settings.orc.filter_push_down && key_condition && !sarg)
+    if (format_settings.orc.filter_push_down && key_condition && !sargs)
+        sargs = buildORCSearchArgument(*key_condition, getPort().getHeader(), file_reader->getType(), format_settings);
+
+    selected_stripes = calculateSelectedStripes();
+    read_iterator = 0;
+    prefetch_iterator = 0;
+
+    if (use_prefetch)
+        prefetchStripes();
+}
+
+void NativeORCBlockInputFormat::prefetchStripes()
+{
+    if (prefetch_iterator >= selected_stripes.size())
+        return;
+
+    size_t total_stripe_size = 0;
+    std::vector<uint32_t> stripes;
+    while (prefetch_iterator < selected_stripes.size() && total_stripe_size < min_bytes_for_seek)
     {
-        sarg = buildORCSearchArgument(*key_condition, getPort().getHeader(), file_reader->getType(), format_settings);
+        int stripe = selected_stripes[prefetch_iterator];
+        stripes.push_back(stripe);
+
+        total_stripe_size += file_reader->getStripe(stripe)->getLength();
+        ++prefetch_iterator;
```

Comment:
Is it ok that we don't update anyhow `read_iterator` here?

Reply:
Yes. `read_iterator` is updated once after `createRowReader` is finished, and `prefetch_iterator` is updated after  single/multiple stripes are prefetched.  And `prefetch_iterator` is always larger than `read_iterator` to make sure prefetched cache could be used while reading from `RowReader.